src/registration.rs

```diff
@@ -14,84 +14,76 @@ use self::hyper::header::Connection;
+            certificate_manager: certificate_manager,
+            top_level_domain: top_level_domain,
+            registration_endpoint: format!("{}/register", registration_endpoint),
+            dns_api_endpoint: dns_api_endpoint,
         }
     }
 
-    pub fn start(&self, endpoint_url: String,
-                 iface: Option<String>,
-                 domain: String,
-                 tunnel: &Option<Tunnel>,
-                 box_port: u16,
-                 dns_api_endpoint: String,
-                 certificate_manager: CertificateManager) {
-        info!("Starting registration with {}", endpoint_url);
-        let endpoint_url = format!("{}/register", endpoint_url);
-
-        let ip_addr = self.get_ip_addr(&iface);
-        if ip_addr == None {
-            // TODO: retry later, in case we're racing with the network
-            // configuration.
-            return;
-        }
-
-        info!("Got ip address: {}", ip_addr.clone().unwrap());
-
-        let certificate_record_result = certificate_manager
-                                            .get_or_generate_self_signed_certificate(&self.box_name);
+    fn get_fingerprint(&self) -> String {
+        self.certificate_manager.get_box_certificate()
+                                .unwrap()
+                                .get_certificate_fingerprint()
+    }
 
-        let (domain, client_fingerprint) = if let Ok(certificate_record) = certificate_record_result {
-            let self_signed_cert_fingerprint = certificate_record.get_certificate_fingerprint();
-            let fingerprint_domain = format!("{}.{}", self_signed_cert_fingerprint, domain);
+    fn get_common_name(&self) -> String {
+        format!("{}.{}", self.get_fingerprint(), self.top_level_domain)
+    }
 
-            info!("Using {}", fingerprint_domain);
-            (fingerprint_domain, self_signed_cert_fingerprint)
-        } else {
-            panic!("Could not get or generate self signed certificate - registration will always fail: {}", certificate_record_result.err().unwrap().description());
-        };
+    pub fn get_local_dns_name(&self) -> String {
+        format!("local.{}", self.get_common_name())
+    }
 
-        let tunnel_url = if let Some(ref tunnel) = *tunnel {
-            tunnel.get_remote_hostname()
-        } else {
-            None
-        };
+    pub fn get_remote_dns_name(&self) -> String {
+        format!("remote.{}", self.get_common_name())
+    }
 
+    fn register_with_registration_server(&self, ip_addr: String, http_scheme: &str, box_port: u16, tunnel_enabled: bool) -> () {
         let message = json!({
-            local_address: format!("local.{}:{}", domain, box_port),
-            tunnel_url: tunnel_url
+            local_url: format!("{}://{}:{}", http_scheme, self.get_local_dns_name(), box_port),
+            tunnel_url: if tunnel_enabled {
+                Some(format!("{}://{}", http_scheme, self.get_remote_dns_name()))
+            } else {
+                None
```

Comment:
Just to double check, in this case the JSON object won' have a field like `"tunnel_url": ""`, right?